text/0000-pbft-consensus.md

```diff
@@ -0,0 +1,480 @@
+  repeated PbftMessage pre_prepare_messages = 3;
+}
+```
+
+## Algorithm Operation
+The PBFT algorithm will operate within the framework described by the Consensus
+API [[3]](#references). The `start` method contains an event loop which
+handles all incoming messages, in the form of `Update`s. The most important
+form of `Update` to the functionality of PBFT is `Update::PeerMessage`, but
+other updates like `BlockNew`, `BlockCommit`, `BlockValid`, and `Shutdown` are
+considered.
+
+### Peer Messages
+When a message arrives from a peer, it must be interrogated for its type, and
+then the system must create a corresponding language-specific object of that
+message type. This is made easier by the fact that all consensus messages are
+protobuf-serialized. Generally, once a message is converted into an appropriate
+object, it needs to be checked for content to make sure everything is
+legitimate. Some of these checks are performed by the validator (checking that
+a message's signature is valid), but some (making sure messages match) need to
+be handled by the PBFT consensus engine.
+
+### Node Information Storage
+Nodes will keep track of the following information:
++ Their own id.
+
++ Log of every peer message that has been sent to it (used to determine if
+  nodes have received enough matching messages to proceed to the next stage of
+  the algorithm, can be [garbage collected](#garbage-collection) every so
+  often).
+
++ List of their connected peers.
+
++ Which step of the algorithm they're on.
+
+### Predicates
+In order to keep the algorithm explanation below concise, we'll define some
+predicates here.
+
++ `prepared` is true for the current node if the following messages are
+  present in its log:
+  + The original `BlockNew` message
+  + A `pre_prepare` message matching the original message (in the current view)
+  + `2f` matching `prepare` messages from different nodes that match
+    `pre_prepare` message above
+
++ `committed` is true if for the current node:
+  + `prepared` is true
+  + This node has accepted `2f + 1` `commit` messages, including its own
+
+### Normal Case Algorithm Operation
+#### Explanation of Message Types
++ `pre_prepare`: Sent from primary node to all nodes in the network, notifying
+them that a new message (`BlockNew`) has been received from the validator.
+
++ `prepare`: Sent from every node to every other node; used as verification of
+the pre-prepare message.
+
++ `commit`: Sent from every node to every other node; used to determine if there
+is consensus that we should indeed commit the block contained in the original
+message.
+
++ `commit_final`: Sent from all nodes back to the primary node; used as
+confirmation that the block should be committed.
+
++ `checkpoint`: Sent by any node that has completed `checkpoint_period`
+`commit_final` messages.
+
++ `view_change`: Sent by any node that suspects that the primary node is faulty.
+
++ `new_view`: Sent by any node that receives `2f` `view_change` messages.
```

Comment:
Should this 2f be (2f+1) for new_view in case f view_changes are byzantine?

Reply:
Yes, this should definitely be `2f + 1` (including the message this node sent to itself)